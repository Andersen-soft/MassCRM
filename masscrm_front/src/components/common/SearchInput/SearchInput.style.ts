import { createStyles, makeStyles } from '@material-ui/core';

export const searchStyle = makeStyles(() =>
  createStyles({
    searchIcon: {
      width: '16px',
      height: '16px',
      position: 'absolute',
      top: '30%',
      right: '6%',
      color: '#69738F',
      '&:focused': {
        display: 'none'
      }
    },
    searchWrap: {
      display: 'flex',
      position: 'relative'
    },
    search: {
      width: '100%',
      '& .MuiAutocomplete-inputRoot .MuiAutocomplete-input': {
        padding: 0
      },
      '& .Mui-focused .search-icon': {
        visibility: 'hidden'
      },
      '& .Mui-disabled': {
        borderColor: 'rgba(225, 229, 237, 0.5)',
        color: 'rgba(225, 229, 237, 0.5)'
      },
      '& .MuiIconButton-root': {
        background: '#fff',
        zIndex: 1
      },
      '& .Mui-disabled:hover ': {
        borderColor: 'rgba(225, 229, 237, 0.5)'
      },
      '& .Mui-error': {
        borderColor: '#FD5757'
      },
      '& .MuiSvgIcon-fontSizeSmall': {
        fontSize: '1rem'
      }
    },
    disabledIcon: {
      color: 'rgba(225, 229, 237, 0.5)'
    },
    required: {
      '& .MuiInputLabel-outlined::before': {
        content: '"*"'
      }
    }
  })
);
