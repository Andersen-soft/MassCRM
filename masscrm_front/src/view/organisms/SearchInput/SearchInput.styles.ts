import { createStyles, makeStyles } from '@material-ui/core';

export const useStyles = makeStyles(() =>
  createStyles({
    searchIcon: {
      width: '16px',
      height: '16px',
      position: 'absolute',
      top: '30%',
      right: '6%',
      color: '#69738F',
      '&:focused': {
        display: 'none'
      }
    },
    searchWrap: {
      display: 'flex',
      flexDirection: 'column',
      position: 'relative'
    },
    search: {
      width: '230px',
      '& .MuiAutocomplete-inputRoot .MuiAutocomplete-input': {
        position: 'relative',
        padding: 0
      },
      '& .Mui-focused .search-icon': {
        visibility: 'hidden'
      },
      '& .Mui-disabled': {
        borderColor: 'rgba(225, 229, 237, 0.5)',
        color: 'rgba(225, 229, 237, 0.5)'
      },
      '& .MuiIconButton-root': {
        background: '#fff',
        zIndex: 1
      },
      '& .Mui-disabled:hover ': {
        borderColor: 'rgba(225, 229, 237, 0.5)'
      },
      '& .Mui-error': {
        borderColor: '#FD5757'
      },
      '& .MuiSvgIcon-fontSizeSmall': {
        fontSize: '1rem'
      }
    },
    disabledIcon: {
      color: 'rgba(225, 229, 237, 0.5)'
    },
    required: {
      '& .MuiInputLabel-outlined::before': {
        content: '"*"'
      }
    },
    paper: {
      position: 'absolute',
      top: '42px',
      width: '230px'
    },
    addButtonBlock: {
      position: 'absolute',
      bottom: '-42px',
      width: '230px',
      background: '#fff',
      zIndex: 2,
      display: 'flex',
      justifyContent: 'center',
      alignItems: 'center',
      height: '42px',
      padding: '5px',
      boxShadow: '0 5px 15px rgba(5, 31, 43, 0.17)',
      borderRadius: '4px',
      boxSizing: 'border-box'
    },
    sendButtonBlock: {
      textAlign: 'center',
      marginTop: '30px'
    }
  })
);
