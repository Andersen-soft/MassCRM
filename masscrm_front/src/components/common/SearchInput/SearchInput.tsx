import React, {
  ChangeEvent,
  FC,
  useMemo,
  useCallback,
  useState,
  useEffect
} from 'react';
import { FormControl, TextField, Box } from '@material-ui/core';
import { Autocomplete } from '@material-ui/lab';
import { Search } from '@material-ui/icons';
import { inputStyle } from 'src/styles/CommonInput.style';
import { ISearchInputProps, ISearchItem } from './interfaces';
import { searchStyle } from './SearchInput.style';

export const SearchInput: FC<ISearchInputProps> = ({
  width,
  placeholder,
  onChange,
  onSelect,
  items,
  value,
  name,
  errorMessage,
  required,
  disabled,
  className,
  freeSolo = true,
  onBlur,
  onFocus,
  multi
}) => {
  const style = inputStyle();
  const classes = searchStyle();
  const rootClasses = useMemo(
    () => ({
      root: `${style.input} ${multi ? style.multiInputForm : style.soloInput} ${
        classes.search
      } ${required && classes.required} ${className}`
    }),
    [className, classes.required, classes.search, required, style.input]
  );

  const [valueState, setValueState] = useState<
    string | ISearchItem | (string | ISearchItem)[] | null
  >(value || '');

  const [inputValueState, setInputValueState] = useState<
    string | ISearchItem | (string | ISearchItem)[] | null
  >(value || '');

  useEffect(() => {
    if (!value) {
      setValueState('');
      setInputValueState('');
    }
    if (multi && Array.isArray(value) && !value.length) {
      setValueState([]);
      setInputValueState('');
    }
  }, [value]);

  const handleChange = (event: ChangeEvent<{}>, val: string) => {
    setInputValueState(val);
    !multi && onChange(val);
  };

  const handleChangeInput = ({
    target
  }: React.ChangeEvent<HTMLInputElement>) => {
    !multi && onChange(target.value);
  };

  const handleSelect = (
    event: ChangeEvent<{}>,
    val: string | ISearchItem | (string | ISearchItem)[] | null
  ) => {
    setValueState(val);
    onChange(val);
    if (onSelect) {
      onSelect(val);
    }
  };

  const handleGetOptionSelected = useCallback(
    (option: string | ISearchItem, currentValue: string | ISearchItem) =>
      typeof option === 'object' && typeof currentValue === 'object'
        ? option.key === currentValue.key
        : option === currentValue,
    []
  );

  const handleGetOptionLabel = useCallback(
    (option: string | ISearchItem) =>
      typeof option === 'string' ? option : option.label,
    []
  );

  return (
    <>
      <FormControl style={{ width }} variant='outlined' classes={rootClasses}>
        <Autocomplete
          id={name}
          options={items}
          multiple={multi}
          limitTags={1}
          size='small'
          freeSolo={freeSolo}
          value={Array.isArray(value) && value.length ? value : valueState}
          inputValue={inputValueState as string}
          onInputChange={handleChange}
          onChange={handleSelect}
          getOptionSelected={handleGetOptionSelected}
          getOptionLabel={handleGetOptionLabel}
          disabled={disabled}
          onBlur={onBlur}
          onFocus={onFocus}
          blurOnSelect={!multi}
          openOnFocus={!multi}
          disableCloseOnSelect={multi}
          renderInput={params => (
            <div className={classes.searchWrap}>
              <TextField
                {...params}
                id={name}
                name={name}
                label={placeholder}
                variant='outlined'
                error={!!errorMessage}
                onChange={handleChangeInput}
              />
              <Search
                className={`${classes.searchIcon} search-icon ${disabled &&
                  classes.disabledIcon}`}
              />
            </div>
          )}
        />
      </FormControl>
      {errorMessage && <Box className={style.error}>{errorMessage}</Box>}
    </>
  );
};
