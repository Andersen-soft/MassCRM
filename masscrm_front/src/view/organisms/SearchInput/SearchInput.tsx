import React, {
  FC,
  useMemo,
  useCallback,
  useState,
  useEffect,
  ChangeEvent
} from 'react';
import { FormControl, TextField, Box } from '@material-ui/core';
import { Autocomplete } from '@material-ui/lab';
import { Search } from '@material-ui/icons';
import { CommonButton } from 'src/view/atoms';
import { ISearchItem } from 'src/interfaces';
import { commonInputStyles } from 'src/styles';
import { AddDialog } from './components';
import { useStyles } from './SearchInput.styles';

interface IProps {
  width?: string;
  placeholder: string;
  items: (string | ISearchItem)[];
  onChange: Function;
  onSelect?: Function;
  value?: string | ISearchItem;
  name?: string;
  errorMessage?: string;
  required?: boolean;
  disabled?: boolean;
  className?: string;
  inputValue?: string;
  freeSolo?: boolean;
  onBlur?: () => void;
  onFocus?: () => void;
  multi?: boolean;
  autoFocus?: string;
  createNewIndustry?: Function;
  isScrollForm?: boolean;
  isFieldMatching?: boolean;
}

export const SearchInput: FC<IProps> = ({
  width,
  placeholder,
  onChange,
  onSelect,
  items,
  value,
  name,
  errorMessage,
  required,
  disabled,
  className = '',
  freeSolo = true,
  onBlur,
  onFocus,
  multi,
  inputValue,
  autoFocus,
  isScrollForm,
  isFieldMatching
}) => {
  const commonInputClasses = commonInputStyles();
  const styles = useStyles();

  const rootClasses = useMemo(
    () => ({
      root: `${commonInputClasses.input} ${styles.search} 
      ${
        multi ? commonInputClasses.multiInputForm : commonInputClasses.soloInput
      } 
      ${required ? styles.required : ''} ${className}`
    }),
    [
      className,
      styles.required,
      styles.search,
      styles.paper,
      required,
      commonInputClasses.input
    ]
  );

  const isIndustry: boolean = name === 'industry';

  const [valueState, setValueState] = useState<
    string | ISearchItem | (string | ISearchItem)[] | null
  >(value || '');

  const [inputValueState, setInputValueState] = useState<
    string | ISearchItem | (string | ISearchItem)[] | null
  >(value || '');

  const [showButton, setShowButton] = useState<boolean>(false);
  const [openDialog, setOpenDialog] = useState<boolean>(false);

  const handleSetOpen = () => setShowButton(prev => !prev);

  useEffect(() => {
    if (!value) {
      setValueState('');
      setInputValueState('');
    }
    if (value && !multi) {
      setInputValueState(value);
    }
    if (multi && Array.isArray(value) && !value.length) {
      setValueState([]);
      setInputValueState('');
    }
  }, [value]);

  const handleChange = (_: ChangeEvent<{}>, val: string) => {
    setInputValueState(val);
    !multi && onChange(val);
  };

  const handleChangeInput = ({ target }: ChangeEvent<HTMLInputElement>) => {
    !multi && onChange(target.value);
  };

  const handleSelect = (
    _: ChangeEvent<{}>,
    val: string | ISearchItem | (string | ISearchItem)[] | null
  ) => {
    setValueState(val);
    onChange(val);
    if (onSelect) {
      onSelect(val);
    }
  };

  const handleGetOptionSelected = useCallback(
    (option: string | ISearchItem, currentValue: string | ISearchItem) =>
      typeof option === 'object' && typeof currentValue === 'object'
        ? option.key === currentValue.key
        : option === currentValue,
    []
  );

  const handleGetOptionLabel = useCallback(
    (option: string | ISearchItem) =>
      typeof option === 'string' ? option : option.label,
    []
  );

  const handleSetOpenDialog = () => setOpenDialog(prev => !prev);

  const handleCreateNewIndustry = useCallback(
    (industryValue: string[]) => {
      onChange(industryValue);
    },
    [valueState, value]
  );

  const getInputValue = () => {
    if (inputValue) {
      return inputValue;
    }
    if (Array.isArray(value) || isFieldMatching) {
      return '';
    }
    return inputValueState as string;
  };

  return (
    <>
      <FormControl style={{ width }} variant='outlined' classes={rootClasses}>
        <Autocomplete
          id={name}
          classes={{ paper: isIndustry ? styles.paper : '' }}
          options={items}
          multiple={multi}
          limitTags={1}
          size='small'
          freeSolo={freeSolo}
          value={Array.isArray(value) && value.length ? value : valueState}
          inputValue={getInputValue()}
          onInputChange={handleChange}
          onChange={handleSelect}
          getOptionSelected={handleGetOptionSelected}
          getOptionLabel={handleGetOptionLabel}
          disabled={disabled}
          onBlur={onBlur}
          onFocus={onFocus}
          onClose={handleSetOpen}
          onOpen={handleSetOpen}
          blurOnSelect={!multi}
          filterOptions={options => options}
          openOnFocus={!multi}
          disableCloseOnSelect={multi}
          renderInput={params => (
            <div className={styles.searchWrap}>
              <TextField
                {...params}
                id={name}
                name={name}
                label={placeholder}
                variant='outlined'
                error={!!errorMessage}
                helperText={isScrollForm && errorMessage}
                onChange={handleChangeInput}
                autoFocus={!!(autoFocus && autoFocus === name)}
              />
              {isIndustry && showButton && (
                <div className={styles.addButtonBlock}>
                  <CommonButton
                    text='Add new industry'
                    color='yellow'
                    onClickHandler={handleSetOpenDialog}
                  />
                </div>
              )}
              <Search
                className={`${styles.searchIcon} search-icon ${disabled &&
                  styles.disabledIcon}`}
              />
            </div>
          )}
        />
        {isIndustry && (
          <AddDialog
            name={name}
            openDialog={openDialog}
            handleSetOpenDialog={handleSetOpenDialog}
            placeholder={placeholder}
            className={rootClasses}
            handleSubmitAction={handleCreateNewIndustry}
            industries={items as string[]}
            existValue={(value as (string | ISearchItem)[] | undefined) || []}
          />
        )}
      </FormControl>
      {errorMessage && !isScrollForm && (
        <Box className={commonInputClasses.error}>{errorMessage}</Box>
      )}
    </>
  );
};
