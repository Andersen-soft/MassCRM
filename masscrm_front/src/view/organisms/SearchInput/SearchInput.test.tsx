import * as React from 'react';
import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import { SearchInput } from './SearchInput';

describe('SearchInput', () => {
  const mockCallBack = jest.fn();

  const searchInput = shallow(
    <SearchInput
      onChange={mockCallBack}
      items={['A', 'B', 'C', 'D', 'E']}
      placeholder='test'
    />
  );

  it('NumericRangeFilter should render correctly', () => {
    expect(shallowToJson(searchInput)).toMatchSnapshot();
  });
});
