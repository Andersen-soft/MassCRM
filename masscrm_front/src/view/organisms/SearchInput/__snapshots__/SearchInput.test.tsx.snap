// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchInput NumericRangeFilter should render correctly 1`] = `
<Fragment>
  <WithStyles(ForwardRef(FormControl))
    classes={
      Object {
        "root": "makeStyles-input-1 makeStyles-search-14 
      makeStyles-soloInput-2 
       ",
      }
    }
    style={
      Object {
        "width": undefined,
      }
    }
    variant="outlined"
  >
    <WithStyles(ForwardRef(Autocomplete))
      blurOnSelect={true}
      classes={
        Object {
          "paper": "",
        }
      }
      filterOptions={[Function]}
      freeSolo={true}
      getOptionLabel={[Function]}
      getOptionSelected={[Function]}
      inputValue=""
      limitTags={1}
      onChange={[Function]}
      onClose={[Function]}
      onInputChange={[Function]}
      onOpen={[Function]}
      openOnFocus={true}
      options={
        Array [
          "A",
          "B",
          "C",
          "D",
          "E",
        ]
      }
      renderInput={[Function]}
      size="small"
      value=""
    />
  </WithStyles(ForwardRef(FormControl))>
</Fragment>
`;
